### ISSUE TITLE
Use Intel's ISA-L_crypto library for computing checksums label:"Type: Performance"   

### ISSUE THREAD

**Issue by jkool702** [openzfs/zfs#14940]
### Describe the feature would like to see added to OpenZFS

I would like to see [Intel's ISA-L_crypto library](https://github.com/intel/isa-l_crypto) utilized for computing checksums ( sha1, sha256, sha512, md5, ... ). This would be utilized at least on Intel CPUs....it may be worthwhile (or required) to continue to support openssl for those using AMD CPUs.

### How will this feature improve OpenZFS?

It has the potential to massively improve performance, especially for people using Intel CPU's that support AVX2 / AVX-512 but ***not*** the SHA-NI instructions (which, I imagine, makes up a decent percentage of the OpenZFS user base, being that this makes up most of the CPUs intel has made in the last decade)

See below for quantatitive info on how much improvent ISA-L_crypto could bring for checksumming. In short: ISA-L_crypto is optimized enough that on basically every modern intel CPU (broadwell and newer) it would ensure that computing checksums (which ZFS does sort of a lot of) is virtually NEVER the "weakest link" that is slowing everything else down and reducing I/O and/or increasing CPU usage.

***

### Additional context

Everything below this point consists of testing results (both from Intel and on my personal machine) in an attempt to make an educated guess of how much switching to ISA-L for checksums would speed up zfs.

TL;DR: for broadwell and newer Intel CPU's, the actual checksums will potentially be computed anywhere between 4-15x faster (for chips with AVX2 and/or AVX-512 but not SHA-NI). This translates into real-world zfs  performance / i/o -speed increasing perhaps by 1/3 (+30-35%).

### Intel's ISA-L Performance Report

See this (admittedly older) [Intel ISA-L performance report](https://01.org/sites/default/files/documentation/intel_isa-l_2.19_performance_report_0_0.pdf) for the type of performance you can expect using ISA-L crypto for hashing/checksumming on Intel CPU's. Below briefly summarizes the important points from the performance report that relate to checksumming:

The increase in performance/efficiency is substantial for basically all intel processors produced in the last decade (broadwell and newer):
* {skylake-->cannon lake} xeons and core-x CPUs + {cannon lake} CPU's (that have AVX-512, but not SHA-NI): potentially an order of magnitude (10x) improvement
* {broadwell-->comet lake} mainstream CPUs (that have AVX2 but not SHA-NI): potentially 5x improvement
* {rocket lake and newer} CPUs that have SHA-NI and AVX-512: potentially 2x improvement (note: this comes from the ISA-L_cryptio github page, not the performance report.

For CPU's with AVX512 but not SHA-NI (page 7), the **per core** performance (at 2.5 GHz) is:
* \>2 G/s for sha512, 
* almost 3 GB/s for sha256, 
* \>5 GB/s for sha1, and 
* nearly 10 GB/s for md5.
Compared to 2017 openssl performance, these are 10-20x faster.

For CPU's with AVX2 but not SHA-NI (page 10), the **per core** performance (at 2.2 GHz) is: 
* \>600 MB/s for sha512, 
* \>800 MB/s for sha256, 
* almost 2 GB/s for sha1, and 
* \~3.5 GB/s for md5. 
Compared to 2017 openssl performance, these are 10-20x faster. With these speeds an older quad-core laptop broadwell CPU could checksum at 3.2 GB/s using SHA256 (or 7-8 GB/s with SHA1), fast enough to saturate the fastest NVME SSDs (excluding a few high-end PCI-v4 ones, which a broadwell system cant use anyways).

### ISA-L_crypto vs openssl: performance testing on my personal machine (i9-7940x)

My personal machine has an (overclocked) 14-core i9-7940x, which falls into the "has AXV-512 but not SHA-NI category. It runs AVX512 stuff at a (hard to keep cool) 3.6 GHz and non-AVX stuff at 4.2-4.4 GHz. It is also paired with 128 gb of exceptionally quick ddr4 RAM (3200 MHz @ 14/14/14/34 + fine-tuned secondary timings).

Below are the results from compiling ISA-L_crypto on this machine and running `make perf`. These have been filtered down to ones related to checksumming and that compare ISA-L to openssl.

```
sha1_openssl_warm: runtime =    1199509 usecs, bandwidth 1250 MB in 1.1995 sec = 1092.71 MB/s
multibinary_sha1_warm: runtime =     160504 usecs, bandwidth 1250 MB in 0.1605 sec = 8166.28 MB/s

md5_openssl_warm: runtime =    1637713 usecs, bandwidth 1250 MB in 1.6377 sec = 800.34 MB/s
multibinary_md5_warm: runtime =     109056 usecs, bandwidth 1250 MB in 0.1091 sec = 12018.78 MB/s

sha256_openssl_warm: runtime =    1012035 usecs, bandwidth 500 MB in 1.0120 sec = 518.05 MB/s
multibinary_sha256_warm: runtime =     119605 usecs, bandwidth 500 MB in 0.1196 sec = 4383.50 MB/s

sha512_openssl_warm: runtime =     181803 usecs, bandwidth 125 MB in 0.1818 sec = 720.96 MB/s
multibinary_sha512_warm: runtime =      37621 usecs, bandwidth 125 MB in 0.0376 sec = 3484.01 MB/s

sm3_openssl_warm: runtime =    4608345 usecs, bandwidth 1250 MB in 4.6083 sec = 284.42 MB/s
multibinary_sm3_warm: runtime =     399529 usecs, bandwidth 1250 MB in 0.3995 sec = 3280.66 MB/s
```

After adjusting for the cloickspeed difference, these numbers agree fairly well with what is presented in Intel's ISA-L Performance Report

### Is checksum computation actually slowing down zfs???

To test this I used my [ZFSEncryption_SpeedTest](https://github.com/jkool702/zfsEncryption_SpeedTest) code. This code creates a zfs pool + filesystem on a ram disk (on a BRD block device is possible, otherwise on a tmpfs) and then tests it using `fio`. I configured this to test with sha256 checksumming and then without checksumming for a mostly read workload, a mostly write workload, and a evenly mixed workload. Results are shown below

    Checksum Type:           sha256        sha256          off            off
    Block Size (KB):         128           4               128            4
    
    10% READ / 90% WRITE:    204MiB/s      36.9MiB/s       287MiB/s       37.7MiB/s  |
    50% READ / 50% WRITE:    1484MiB/s     283MiB/s        2021MiB/s      264MiB/s   |--- READ SPEED
    90% READ / 10% WRITE:    5035MiB/s     1134MiB/s       5490MiB/s      1132MiB/s  |
    
    10% READ / 90% WRITE:    1631          9450            2292           9650       |
    50% READ / 50% WRITE:    11.9k         72.4k           16.2k          67.5k      |--- READ IOPS
    90% READ / 10% WRITE:    40.3k         290k            43.9k          290k       |
    
    90% READ / 10% WRITE:    559MiB/s      126MiB/s        609MiB/s       126MiB/s   |
    50% READ / 50% WRITE:    1484MiB/s     283MiB/s        2022MiB/s      263MiB/s   |--- WRITE SPEED
    10% READ / 90% WRITE:    1843MiB/s     334MiB/s        2591MiB/s      341MiB/s   |
    
    90% READ / 10% WRITE:    4470          32.2k           4874           32.2k      |
    50% READ / 50% WRITE:    11.9k         72.4k           16.2k          67.4k      |--- WRITE IOPS
    10% READ / 90% WRITE:    14.7k         85.4k           20.7k          87.2k      |
    
    
The improvement in BOTH read speed and write speed (at 128 KB blocksize) was as follows:
    
    90% READ / 10% WRITE:    9%
    50% READ / 50% WRITE:    36%
    10% READ / 90% WRITE:    40.6%
	
This suggests that, were checksum computation to become massively more efficient, zfs i/o speed might increase by up to ~1/3 for workloads with at least a moderate amount of write i/o. This will most likely be accompanied by lower overall CPU usage.
